File: leet-code/78. Subsets/Subsets78.java
import java.util.ArrayList;
import java.util.List;

public class Subsets78 {
    public List<List<Integer>> subsets(int[] nums) {
        return dfs(new ArrayList<>(), new ArrayList<>(), nums, 0);
    }

    private List<List<Integer>> dfs(List<List<Integer>> subsets, List<Integer> subset, int[] nums, int idx) {
        subsets.add(new ArrayList<>(subset));
        for (int i = idx; i < nums.length; i++) {
            subset.add(nums[i]);
            dfs(subsets, subset, nums, i + 1);
            subset.remove(subset.size() - 1);
        }
        return subsets;
    }
}

File: leet-code/78. Subsets/Subsets78.kt
package com.example

class Subsets78 {
    fun subsets(nums: IntArray): List<List<Int>> {
        val result = mutableListOf<List<Int>>()
        val subset = mutableListOf<Int>()

        fun dfs(idx: Int) {
            result.add(subset.toList())
            for (i in idx..nums.size - 1) {
                subset.add(nums[i])
                dfs(i + 1)
                subset.removeLast()
            }
        }

        dfs(0)
        return result
    }
}
